### PR TITLE
Test/sync

### PR DESCRIPTION
**Pull request by haydenyoung**

### CODE REVIEW

File: test/sync.test.js [hunk 1]
```diff
@@ -332,6 +338,8 @@ describe('Sync protocol', function () {
       await log1.append('hello9')
       expectedEntry2 = await log1.append('hello10')
 
+      expectedEntry = expectedEntry2
+
```
Comment (haadcode) [inline test/sync.test.js:400]: This seems extraneous as it’s not used in the code below, right?

File: src/sync.js [hunk 2]
```diff
@@ -3,6 +3,31 @@ import PQueue from 'p-queue'
...
+ * - There is no guarantee that all heads will be received.
+ * - There is no guarantee that heads will reflect the latest heads on the remote peer.
+ * - Calling Sync.add is neither synchronous nor deterministic. Calling Sync.add with async/await only guarantees that the message is published. It does not guarantee when the message will be received or even that the message is recieved at all.
+ * - Determining whether a peer has received a message can be handled by registering a callback using onSynced. However, there is no guarantee that every log entry will be received and onSynced can not be used to determine whether every head is received.
+ * - The only guarantee that Sync provides is that heads will eventually reach consistency between all peers.
+ */
+
+/**
+ * Creates a Sync instance for sychronizing logs between multiple peers.
+ * @param {Object} options One or more options for configurating Sync.
```
Comment (haadcode) [inline src/sync.js]: I think the params here are not options (as in optionals) but rather required params.

File: src/sync.js [hunk 3]
```diff
@@ -3,6 +3,31 @@ import PQueue from 'p-queue'
...
+ * @param {IPFS} options.ipfs An IPFS instance
+ * @param {Log} options.log A Log instance.
```
Comment (haadcode) [inline src/sync.js]: “The Log instance to sync”

File: src/sync.js [hunk 4]
```diff
@@ -3,6 +3,31 @@ import PQueue from 'p-queue'
...
+ * @param {Object} options.events An event emitter. Defaults to an instance of EventEmitter.
```
Comment (haadcode) [inline src/sync.js]: Let’s add to the description which events this emitter will emit and when.

File: src/sync.js [hunk 5]
```diff
@@ -3,6 +3,31 @@ import PQueue from 'p-queue'
...
+ * @param {Function} options.onSynced A callback function for determining whether an entry has been receieved.
```
Comment (haadcode) [inline src/sync.js]: “A callback that is called after the peer has received heads from another peer”

File: src/sync.js [hunk 6]
```diff
@@ -3,6 +3,31 @@ import PQueue from 'p-queue'
...
+ * @param {Boolean} options.start True if synchronization should happen automatically, false otherwise. Defaults to true.
```
Comment (haadcode) [inline src/sync.js]: “If sync should *start* automatically”

File: test/sync.test.js (hunk 1 above)
Comment (haydenyoung) [reply test/sync.test.js:400]: expectedEntry is needed for determining whether syncedEventFired is true, here: 

```
      const onSynced = async (bytes) => {
        syncedHead = await Entry.decode(bytes)
        if (expectedEntry) {
          syncedEventFired = expectedEntry.hash === syncedHead.hash
        }
      }
```

This is probably a bit messy and should be refactored at a later stage.

Comment (haydenyoung) [reply test/sync.test.js:400]: Refactored.

**Pull request merged**